### PR TITLE
Add a `ScopedVariable` trait

### PR DESCRIPTION
**Pull request by Laurens-W**
## What's changed?

Added a trait to pull together the scope and identifier of `NamedVariable`s and added testcases to showcase its usage

## Anyone you would like to review specifically?

@sambsnyd @timtebeek 

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [x] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

File: src/main/java/org/openrewrite/java/migrate/trait/ScopedVariable.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import lombok.RequiredArgsConstructor;
+import lombok.Value;
+import org.jspecify.annotations.Nullable;
+import org.openrewrite.Cursor;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.trait.SimpleTraitMatcher;
+import org.openrewrite.trait.Trait;
+
+@Value
+public class ScopedVariable implements Trait<J.VariableDeclarations.NamedVariable> {
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/java/migrate/trait/ScopedVariable.java:27]: Shouldn't we instead create this trait in the rewrite repo?

File: src/main/java/org/openrewrite/java/migrate/trait/ScopedVariable.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    J.Identifier identifier;
+
+    @RequiredArgsConstructor
+    public static class Matcher extends SimpleTraitMatcher<ScopedVariable> {
+
+        @Override
+        protected @Nullable ScopedVariable test(Cursor cursor) {
+            if (cursor.getValue() instanceof J.VariableDeclarations.NamedVariable) {
+                J.VariableDeclarations.NamedVariable variable = cursor.getValue();
+                return new ScopedVariable(cursor, variable.getDeclaringScope(cursor), variable.getName());
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/java/migrate/trait/ScopedVariable.java:39]: This `getDeclaringScope()` utility method is probably a good starting point. But to me it looks like it misses cases like try-with-resources and exception handlers, for example. Also note that the flow scoping of instanceof patterns is quite different and probably a bad fit. Maybe worth thinking about at least.

File: src/main/java/org/openrewrite/java/migrate/trait/ScopedVariable.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    Cursor cursor;
+    Cursor scope;
+    J.Identifier identifier;
+
+    @RequiredArgsConstructor
+    public static class Matcher extends SimpleTraitMatcher<ScopedVariable> {
+
+        @Override
+        protected @Nullable ScopedVariable test(Cursor cursor) {
+            if (cursor.getValue() instanceof J.VariableDeclarations.NamedVariable) {
```
Comment (amishra-u) [inline src/main/java/org/openrewrite/java/migrate/trait/ScopedVariable.java:37]: We can likely just use variable.getDeclaringScope(cursor) to obtain the variable's scope; using a trait might be overkill.

A scenario where a trait could be useful, suppose I'm at an identifier and need to determine the declaring scope of the variable that this identifier references.

```
void foo() {
   int x = 5;
   if (true) {
      print(x) // find the declaring scope of x at this point,
   }
}
```

**Comment by Laurens-W** [comment]
Continued here
- https://github.com/openrewrite/rewrite/pull/4682

**Pull request closed without merging**